### ISSUE TITLE
BYOM not working with LEO?

### ISSUE THREAD

**Issue by jhmonroe** [brave/brave-browser#42819]
### Description

I've texted with Anthropic/Claude, OpenAI and Perplexity and cannot get Leo to work with any BYOM configs.

I have all the **endpoints** that I use in other programs
https://api.anthropic.com/v1/messages
https://api.openai.com/v1/chat/completions
https://api.perplexity.ai

I tried different model names according to the provider's sites:
https://docs.anthropic.com/en/docs/about-claude/models
https://platform.openai.com/docs/models
https://docs.perplexity.ai/guides/model-cards

I have my **API Keys** for all

What am I missing? Does Leo just not work with outside BYOM so people will pay for Premium?
I've gotten this working fine in Obsidian, Palchat, etc so it can't be user error... can it?

### Steps to reproduce

1. try models
2. ask simple question
3. hit retry over and over
4. never works

### Actual result

![Image](https://github.com/user-attachments/assets/2d214884-25ad-4dad-9ad1-3a45fed8aecf)

### Expected result

![Image](https://github.com/user-attachments/assets/939cf6c5-0c94-4b6d-9345-5e2e05feab40)

### Reproduces how often

Easily reproduced

### Brave version (brave://version info)

Brave is up to date
[Version 1.73.97 Chromium: 131.0.6778.108 (Official Build) (x86_64)](https://brave.com/latest/)

### Channel information

- [x] release (stable)
- [ ] beta
- [ ] nightly

### Reproducibility

- [ ] with Brave Shields disabled
- [ ] with Brave Rewards disabled
- [ ] in the latest version of Chrome

**Comment by jonathansampson** [comment]
@jhmonroe Looking into these issues now. OpenAI is definiately an endpoint we support. From the looks of your screenshot, the API Key appears to be considerably shorter than what I would expect. My testing key, for example, is 151 characters long. Please visit https://platform.openai.com/settings/organization/api-keys and make sure that you are using the correct private/secret key value.

**Comment by jhmonroe** [comment]
Yes it's from that API key page. My keys are only 56 chars and work fine with lots of other apps

Claude uses 156 character

Today OpenAI (with API Key) seems to be working in Leo, but Claude and Perplexity (with API Keys) do not work.
I do know from experience using open-AI compatible models inside Obsidian that the CORS option needs to be enabled.
Perhaps this is the issue Leo is experiencing...

**Comment by jonathansampson** [comment]
I will definitely look more closely at those other endpoints, but wanted to make sure we hadn't regressed with OpenAI. Happy to hear that OpenAI _is indeed working_ for you. I'll look into Anthropic and Perplexity ASAP.

**Comment by jhmonroe** [comment]
> I will definitely look more closely at those other endpoints, but wanted to make sure we hadn't regressed with OpenAI. Happy to hear that OpenAI _is indeed working_ for you. I'll look into Anthropic and Perplexity ASAP.

thank you!! I just added more info about what I tried and what is working not and not working

**Comment by jonathansampson** [comment]
> > I will definitely look more closely at those other endpoints, but wanted to make sure we hadn't regressed with OpenAI. Happy to hear that OpenAI _is indeed working_ for you. I'll look into Anthropic and Perplexity ASAP.
> 
> thank you!! I just added more info about what I tried and what is working and not working

Sounds like OpenAI and Perplexity are both working then. Only Anthropic is failing for you? This would make sense; our initial support under BYOM was for the OpenAI API, which is implemented in large part by various other providers. As such, if an endpoint supports the OpenAI API style, then Brave should be able to support that endpoint already. Anthropic appears to have some minor requirements that are currently unsupported under our model (e.g., the presence of an `anthropic-version` header). That said, we should be able to make a few changes and extend support to Anthropic as well.

**Comment by jhmonroe** [comment]
> > > I will definitely look more closely at those other endpoints, but wanted to make sure we hadn't regressed with OpenAI. Happy to hear that OpenAI _is indeed working_ for you. I'll look into Anthropic and Perplexity ASAP.
> > 
> > 
> > thank you!! I just added more info about what I tried and what is working and not working
> 
> Sounds like OpenAI and Perplexity are both working then. Only Anthropic is failing for you? This would make sense; our initial support under BYOM was for the OpenAI API, which is implemented in large part by various other providers. As such, if an endpoint supports the OpenAI API style, then Brave should be able to support that endpoint already. Anthropic appears to have some minor requirements that are currently unsupported under our model (e.g., the presence of an `anthropic-version` header). That said, we should be able to make a few changes and extend support to Anthropic as well.

Interesting! I guess I assumed Claude was OpenAI compatible since my other apps that use LangChain and OpenAI compatible things can play nice with Claude. But I guess it makes sense it's not OAI compatible since it was started by defectors! ha :-)

**Comment by dhcgn** [comment]
I used Brave with Bring Your Own Model (BYOM) in the past, but now when I tried it again, I got these same error.

"There was a network issue connecting to Leo. Check your connection and try again."
![Image](https://github.com/user-attachments/assets/1658c6db-5d0d-4ddc-885b-83151b066f0c)

Tested on:
-  v1.74.48 
- v 1.76.21 (Nightly)

Api-Keys are working!

**Comment by vorpalhex** [comment]
I can't get this to work with other OpenAI compatible API endpoints either, despite eg Open WebUI working with those same endpoints! 

Is there a way to get better logs here? I'm not seeing anything in the standard log output.

**Comment by jonathansampson** [comment]
> I can't get this to work with other OpenAI compatible API endpoints either, despite eg Open WebUI working with those same endpoints! 
> 
> Is there a way to get better logs here? I'm not seeing anything in the standard log output.

Can you share examples of endpoints and configuration details that aren't working?

**Comment by vorpalhex** [comment]
> > I can't get this to work with other OpenAI compatible API endpoints either, despite eg Open WebUI working with those same endpoints!
> > Is there a way to get better logs here? I'm not seeing anything in the standard log output.
> 
> Can you share examples of endpoints and configuration details that aren't working?

Sure, here's a curl response from my model:

```
HTTP/1.1 200 OK
Content-Length: 1592
Content-Type: application/json; charset=utf-8
Date: Sat, 01 Feb 2025 03:17:02 GMT

{
    "choices": [
        {
            "finish_reason": "stop",
            "index": 0,
            "logprobs": null,
            "message": {
                "content": "Alright, linear regression is like trying to draw a straight line through a bunch of points on a graph. Picture this: you've got some data points that show a relationship between two things—like how much you study and your test scores. Linear regression will help you find the best straight line that predicts your test score based on how many hours you hit the books.\n\nMathematically, it looks like this: y = mx + b. Here, y is what you’re trying to predict (like test scores), x is your input (study hours), m is the slope of the line (how much your score goes up for each extra hour of studying), and b is where the line crosses the y-axis (your score when you study zero hours, which is probably a sad number).\n\nOnce you have that line, you can use it to make predictions. It’s a pretty straightforward way to understand relationships in data, but it can get messy if the data doesn’t actually fit a straight line. Just don’t get too attached to the line; it’s not a commitment.",
                "refusal": null,
                "role": "assistant"
            }
        }
    ],
    "created": 1738379822,
    "id": "chatcmpl-929586aa-3cdf-4a08-bd74-8b2b03d5c8db",
    "model": "iris-dm",
    "object": "chat.completion",
    "service_tier": "default",
    "system_fingerprint": "fp_df52e04a-be1d-4dbe-b22f-1f5c1afe1122",
    "usage": {
        "completion_tokens": 982,
        "completion_tokens_details": {
            "accepted_prediction_tokens": 0,
            "audio_tokens": 0,
            "reasoning_tokens": 0,
            "rejected_prediction_tokens": 0
        },
        "prompt_tokens": 35,
        "prompt_tokens_details": {
            "audio_tokens": 0,
            "cached_tokens": 0
        },
        "total_tokens": 1017
    }
}
```

Server endpoint is configured in Brave as `https://<HOST>/v1/chat/completions` with an appropriate API key. 

I can see the request get to the model and the model responds in a timely fashion (< 1s) .. and Brave just gives me a:

```
There was a network issue connecting to Leo, check your connection and try again.
```

This exact same setup, in the same situation, from the same host works with Open WebUI and other "OpenAI compatible" tools. 

If I could at least get some kind of logs from Brave, I could troubleshoot what's going on..

**Comment by jonathansampson** [comment]
@vorpalhex I suspect the issue in your case is that your endpoint is not _streaming_ chunked responses, but rather sending the entire response all at once. Are you able to modify the endpoint configuration to accept requests for streaming chunks?
